Update navigation to be centered around models

edmundito opened the pull request:
## Change description

This change primarily makes the models and the model overview the center of the navigation in the UI apps.

![image](https://user-images.githubusercontent.com/168664/148616761-5a53b8c8-a400-458d-b1ec-6e347614c9e8.png)

Changes:
- Source schedule is created as "recurring"
- Once the primary source schedule is created, user is redirected to the multiple properties page
- Updates the navigation to show all the models on the side. Records is now tucked away in Platform, and Sources/Group/Destinations list pages are gone
- In nav bar, moved Apps outside of platform and under Dashboard
- Breadcrumbs within the model change from `Type / Topic Name / Tab` to `Model Name / Topic Name / Tab`
- Added breadcrumbs and tabs to model overview to have consistent navigation across all ares
- Spiffied up the navigation bar a bit (minor size and spacing changes)
- Removed navigationModel from global props

## Checklists

### Development

- [x] Application changes have been tested appropriately

### Impact

- [ ] Code follows company security practices and guidelines
- [ ] Security impact of change has been considered
- [ ] Performance impact of change has been considered
- [ ] Possible migration needs considered (model migrations, config migrations, etc.)

Please explain any security, performance, migration, or other impacts if relevant:

### Code review

- [x] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached where applicable.
- [x] Relevant tags have been added to the PR (bug, enhancement, internal, etc.)

edmundito committed:
Update schedule creation after setting mapping to be recurring

edmundito committed:
Navigate to add mutiple properties page after setting schedule for the primary source.
Fix total properties count in properties:list action

edmundito committed:
Hide tabs and only show breadcrumbs if the currently selected page is not part of the tabs

edmundito committed:
Navigate to model overview instead of model edit from model list

edmundito committed:
Update breadcrumbs for destination, group, and property pages to avoid showing the list pages and show instead the model name

edmundito committed:
Show tabs and breadcrumbs in model overview for consistent navigation

edmundito committed:
Take the user to model overview instead of edit after creating a model

edmundito committed:
Update navigation bar to show models
Remove model dropdown from navigation
Remove sources, destinations, groups, properties
Move recofds to Platform section
Improve how the dividers render in the navigation
Adjust margins
Make the expand/collapse icon move

edmundito committed:
Update copy: Add Model -> New Model

bleonard commented:
`Records` (and anything else that's model-specific. is there more?) should be removed from the left menu. That will now be accessible via `model/overview` (See All Records)

bleonard commented on core/src/actions/navigation.ts:
```diff
@@ -190,13 +188,22 @@ export class NavigationList extends OptionallyAuthenticatedAction {
       { type: "link", title: "Export Processors", href: "/exportProcessors" },
     ];
 
+    if (modelId) {
```
I think if my design works, this endpoint would no longer take `modelId` because there is no more concept of "current model" - I noticed there is still several in an `if` block below in `authenticatedConfigNav`. Might some refactoring to combine then here, too.

edmundito committed:
Remove Records nav item

edmundito committed:
Update navigation for config mode
Fix navigation tests
Remove clipboard icon as it's no longer used

edmundito committed:
Update custom model icon to a database instead of cogs

edmundito committed:
Fix UI tests to reflect changes in navigation sidebar

edmundito committed:
Fix PreoprtyTab component test

edmundito committed:
Remove navigationModel
Update ui-config sign-in redirect from records to index page
Update ui-config index CTA from setup records to no CTA

edmundito commented:
Updated to address feedback and I removed the ideas of the current `navigationModel` based on a cookie since it's no longer relevant as the user selects the model

edmundito committed:
Redirect to first model records page on ui config login or setup complete

edmundito committed:
Update ui-config to redirect to first model after sign in or landing in home page

edmundito commented:
Ok, I made all the changes I wanted to make and this should be ready for review. I  resolved a couple of issues in UI config in regards to where to go after completing the setup or signing in. Previously it was going to the record list of the selected model. Now it will redirect the first model.

teallarson reviewed (commented):
This is a nice looking iteration!  

In the future we may want to decide what styling lives on the component/page files and what lives in SCSS modules as there seems to be a mix of it across the UI projects, but I don't see that that needs to be solved here.

pedroslopez commented on core/src/actions/properties.ts:
```diff
@@ -108,7 +108,7 @@ export class PropertiesList extends AuthenticatedAction {
     }
 
-    const total = await Property.scope(null).count();
+    const total = await Property.scope(null).count({ where });
```
Could also use `Property.findAndCountAll` above instead, which will consider the `where` and `includes` for the returned count

pedroslopez commented on ui/ui-components/__tests__/components/tabs.tsx:
```diff
@@ -7,24 +7,28 @@ describe("<PropertyTab />", () => {
+  it("renders the model link properly", async () => {
+    const link = screen.getByRole("link", { name: model.name });
+    expect(link).toHaveAttribute("href", `/model/${model.id}/overview`);
   });
```
Why was this one removed?

edmundito replied:
The breadcrumbs have changed from:

`Properties / {Property Name} / Edit`

To:

`{Model Name} / {Property Name} / Edit`

pedroslopez commented on ui/ui-components/components/Navigation.tsx:
```diff
@@ -205,60 +203,35 @@ export default function Navigation(props) {
-                    />
-                  </Fragment>
+                  <HighlightingNavLink
+                    key={idx}
```
`nav.href` as a key like was being used before is probably more appropriate, unless we don't have unique links now?

edmundito replied:
Make sense, I must have replaced it with the wrong thing when I removed the Fragment.

pedroslopez commented on ui/ui-components/pages/object/[id].tsx:
```diff
@@ -123,21 +121,7 @@ export default function FindObject(props) {
-
-    return `/${tableName}`;
-  }
+  const getListPage = (tableName: string) => `/${tableName}`;
```
I believe this means that in community edition, where we show list pages instead of edit pages for most "tables", it would link to something like `/groups` or `/properties`, which would 404. Let's check what the behavior is for going to `/object/some_group_id` in community edition after this change and ensure it still works. I think some additional change would need to be made to instead take them to the model page or something like that.

edmundito replied:
Ok. I need to rethink how this is done right now because I will need to resolve the model id some other way.

pedroslopez commented on ui/ui-components/pages/setup.tsx:
```diff
@@ -37,10 +35,16 @@ export default function Page(props) {
       "get",
       `/setupSteps`,
-      { modelId: navigationModel.value },
+      { modelId: model?.id },
```
I don't think it's necessary to have this as part of the action input anymore since the model won't be coming from the UI. It can probably be determined somewhere in the action or closer to where it's being used.

edmundito replied:
Agreed. I was working on this while @evantahler was making a change in #2773 but now that it's done and merged this can be removed and moved into the action itself.

pedroslopez commented on ui/ui-components/utils/modelHelper.ts:
```diff
@@ -1,45 +1,6 @@
-  }
-};
-
 export const getModelFromUrlOrCookie = (ctx: NextPageContext) => {
```
Is the cookie part of this still needed? (and the `setModelCookie` function?)

edmundito replied:
Nope, will remove.

pedroslopez reviewed (changes requested):
Nice work!

Mainly some non-blocking questions- only thing I would consider blocking is to check out the `/object/` page in community edition with the change to the function to get the relevant object's list page and ensuring that still works.

edmundito committed:
Use findAndCountAll on property list count

edmundito committed:
Remove `modelHelper` util and usages as they are no longer needed.

edmundito committed:
Update HilightingNavLink key from id to herf

edmundito committed:
Remove modelId from setupSteps action params

edmundito committed:
Fix resolving list page for results in community ui

edmundito committed:
Fix bad GrouparooModel import in setupSteps

pedroslopez commented on core/src/actions/setupSteps.ts:
```diff
@@ -4,13 +4,14 @@ import { AsyncReturnType } from "type-fest";
   description = "List the SetupSteps and their status";
   permission: ActionPermission = { topic: "setupStep", mode: "read" };
   outputExample = {};
-  inputs = { modelId: { required: false } };
```
let's also remove the `modelId` from the `runWithinTransaction` method param types - looks like `{ params }: { params: { modelId?: string } }` is still left over

pedroslopez commented on ui/ui-components/pages/object/[id].tsx:
```diff
@@ -54,10 +54,24 @@ export default function FindObject(props) {
     ) {
-      const listPage = getListPage(response.records[0].tableName.toLowerCase());
-      router.push(listPage);
+      const tableName = response.records[0].tableName.toLowerCase();
```
Note that there's an incoming change in this area here https://github.com/grouparoo/grouparoo/pull/2783

edmundito replied:
Ahh maybe I should apply the change here instead.

pedroslopez commented on core/src/actions/properties.ts:
```diff
@@ -108,7 +108,7 @@ export class PropertiesList extends AuthenticatedAction {
       }
     }
 
-    const total = await Property.scope(null).count();
```
Sorry, I should have been more clear in my suggestion. The tests are failing because `findAndCountAll` returns an object with the rows and the count, so you can be sure the total count refers to whatever you're looking up. My suggestion was to replace `await Property.scope(null).findAll({` on line 71 with `await Property.scope(null).findAndCountAll({` - here's an example on the `exports` action:

https://github.com/grouparoo/grouparoo/blob/c7bce3cac78f48bfc6ace66da041c4c538a95257/core/src/actions/exports.ts#L45

pedroslopez replied:
This works, but now the data is being fetched twice. 

I'm referring to removing this line from here completely and making a change higher up:

https://github.com/grouparoo/grouparoo/blob/041e196e850bf83d12ff6012dfd555d9437b99c0/core/src/actions/properties.ts#L70-L80

This would instead be something like:

```ts
    const { rows, count: total } = await Property.scope(null).findAndCountAll({ 
        include, 
        limit: params.limit, 
        offset: params.offset, 
        order: params.order, 
        where, 
    });

    const properties = rows.filter((p) => 
	params.modelId ? p.source?.modelId === params.modelId : true 
   ); 
```
That way we're grabbing the total and finding in the same call, guaranteeing the same args.
Let me know if you want to have a quick chat

edmundito replied:
Oh of course. 🤦🏼  Despite the name of the function being super clear about what it didn't I didn't think too much into it when I swapped it.

pedroslopez reviewed (commented):
Looking good- I think I know why that test is failing. More info in the comments :D

edmundito committed:
Remove unused params from `setupSteps:list` action

edmundito committed:
Fix total value type in `properties:list` action

edmundito committed:
Update object page to redirect to record edit page on community UI

edmundito committed:
Update model link from edit to overview in import page

edmundito committed:
Only show model tabs in non-community UI

edmundito committed:
Add blockedUiEditions to NavigationItems and remove New Model navigation link from community edition

edmundito committed:
Update links from model edit to model overview page

edmundito committed:
Fix community ui links in model overview

edmundito committed:
Rename navigation ui filter to hideOn to match Link component convention

edmundito commented:
Ok, I pushed a few more changes to fix some issues with the community edition:

- All model links point to the overview page instead of the edit page
- The edit links are removed from the model overview page in the community edition
- Groups point to members in the community edition and rules everywhere else
- Hide the "New Model" link in community edition
- The `/object/{recordId}` page now redirects to the record edit page (implemented the changes from #2783)

Finally, addressed @pedroslopez's comments

pedroslopez commented on core/src/actions/navigation.ts:
```diff
@@ -102,16 +91,45 @@ export class NavigationList extends OptionallyAuthenticatedAction {
+      href: `/model/new`,
+      mainPathSectionIdx: 2,
+      small: hasModels,
+      hideOn: ["community"],
```
Note that you should have access to the ui edition here through `process.env.GROUPAROO_UI_EDITION`, so passing this along might not be necessary- not a huge deal though

edmundito committed:
Exclude new model link from navigation action instead of component

edmundito committed:
Remove redundant count call in properties list action

edmundito commented:
@pedroslopez I fixed the couple of things you mentioned in your review.

pedroslopez reviewed (approved):
Woo! As far as the changes this PR is addressing, looks good to me.

I did notice that sometimes the breadcrumbs show the model capitalized and other times it's not, but I'm not sure why (this was tested in community edition):

Destination exports:
![image](https://user-images.githubusercontent.com/4368928/149224993-4bc49bcc-582e-4819-9d40-e7db876f4abc.png)

Single Record view:
![image](https://user-images.githubusercontent.com/4368928/149225085-7d1cd623-78e3-4618-8b48-3b8dc1b74cf3.png)

I assume this is probably because of showing the ID vs the name in some cases.

Also, this is unrelated but I'm noticing now after this change community edition has no way of running a specific schedule and can now only run all. Added a story for that here https://www.pivotaltracker.com/story/show/180894470.

edmundito committed:
Update links to  and redirect to other pages

evantahler commented on core/src/actions/navigation.ts:
```diff
@@ -12,29 +12,26 @@ type NavigationMode =
-  inputs = {
-    modelId: { required: false },
-  };
+  inputs = {};
```
Note: You can delete them if they aren't in use - not a required property for an Action

evantahler commented on ui/ui-components/__tests__/utils/modelHelper.ts:
```diff
@@ -1,109 +0,0 @@
-import type { NextPageContext } from "next";
```
👋

evantahler commented on ui/ui-components/components/import/List.tsx:
```diff
@@ -13,6 +13,7 @@ import { ErrorHandler } from "../../utils/errorHandler";
 import { ImportRecordPropertiesDiff, ImportGroupsDiff } from "./Diff";
 import StateBadge from "../badges/StateBadge";
 import { capitalize } from "../../utils/languageHelper";
+import { grouparooUiEdition } from "../../utils/uiEdition";
```
Un-used

evantahler reviewed (approved):
This looks great.

Simply because of the number of UI files changed, we'll likely want to do a pre-v0.8 release bug bash, just to make sure that everything still works as expected. (cc @bleonard)

edmundito committed:
Update Record tabs to use model name and load model in record pages

edmundito committed:
Update links in setupSteps to redirect to new model when no model is available

edmundito commented:
@pedroslopez I addressed the couple of issues you mentioned and we should be good to go!

I did notice that we were still redirecting to `/models` in some instances so I made a couple of updates to redirect to either `/` or `/model/new` depending on the context.

edmundito committed:
Code cleanup
